File: aircon-common/src/main/java/com/ironsource/aura/aircon/common/annotations/config/Configs.java
package com.ironsource.aura.aircon.common.annotations.config;

import com.ironsource.aura.aircon.common.utils.Utils;

import java.lang.annotation.Annotation;
import java.util.Arrays;
import java.util.List;

/**
 * Created on 10/20/2018.
 */
public interface Configs {

	List<Class<? extends Annotation>> PRIMITIVES = Arrays.asList(BooleanConfig.class, IntConfig.class, LongConfig.class, FloatConfig.class, StringConfig.class, StringSetConfig.class);
	List<Class<? extends Annotation>> ENUM       = Arrays.asList(IntEnumConfig.class, StringEnumConfig.class);
	List<Class<? extends Annotation>> COMPLEX    = Arrays.asList(TimeConfig.class, JsonConfig.class, ColorConfig.class, UrlConfig.class, TextConfig.class, StyledTextConfig.class);

	List<Class<? extends Annotation>> ALL = Utils.concatLists(PRIMITIVES, ENUM, COMPLEX);
}

File: aircon/src/main/java/com/ironsource/aura/airconkt/config/Configs.kt
package com.ironsource.aura.airconkt.config

fun intConfig(block: ReadWriteConfig<Int, Int>.() -> Unit) = createConfig(block) {
    createConfig(block) {
        PrimitiveConfigDelegate(ConfigSourceResolver.Int, ResourcesResolver.Int)
    }
}

fun longConfig(block: ReadWriteConfig<Long, Long>.() -> Unit) = createConfig(block) {
    PrimitiveConfigDelegate(ConfigSourceResolver.Long, ResourcesResolver.Long)
}

fun floatConfig(block: ReadWriteConfig<Float, Float>.() -> Unit) = createConfig(block) {
    PrimitiveConfigDelegate(ConfigSourceResolver.Float, ResourcesResolver.Float)
}

fun stringConfig(block: ReadWriteConfig<String, String>.() -> Unit) =
        PrimitiveConfigDelegate(ConfigSourceResolver.String, ResourcesResolver.String)

fun booleanConfig(block: ReadWriteConfig<Boolean, Boolean>.() -> Unit) = createConfig(block) {
    PrimitiveConfigDelegate(ConfigSourceResolver.Boolean, ResourcesResolver.Boolean)
}
